go/worker/txnscheduler/worker.go

```diff
@@ -159,36 +173,107 @@ func newWorker(
 	registration *registration.Worker,
+	checkTxEnabled bool,
 ) (*Worker, error) {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+
 	w := &Worker{
-		enabled:      enabled,
-		commonWorker: commonWorker,
-		registration: registration,
-		compute:      compute,
-		runtimes:     make(map[common.Namespace]*committee.Node),
-		quitCh:       make(chan struct{}),
-		initCh:       make(chan struct{}),
-		logger:       logging.GetLogger("worker/txnscheduler"),
+		enabled:        enabled,
+		checkTxEnabled: checkTxEnabled,
+		commonWorker:   commonWorker,
+		registration:   registration,
+		compute:        compute,
+		runtimes:       make(map[common.Namespace]*committee.Node),
+		ctx:            ctx,
+		cancelCtx:      cancelCtx,
+		quitCh:         make(chan struct{}),
+		initCh:         make(chan struct{}),
+		logger:         logging.GetLogger("worker/txnscheduler"),
 	}
 
 	if enabled {
 		if !w.commonWorker.Enabled() {
 			panic("common worker should have been enabled for transaction scheduler")
 		}
 
+		var err error
+
+		// Create the runtime host worker.
+		w.logger.Debug("XXXXXX: Creating runtime host worker...")
+		w.RuntimeHostWorker, err = workerCommon.NewRuntimeHostWorker(commonWorker)
+		if err != nil {
+			return nil, err
+		}
+		w.logger.Debug("XXXXXX: Runtime host worker creation complete.")
+
 		// Use existing gRPC server passed from the node.
 		api.RegisterService(commonWorker.Grpc.Server(), w)
 
 		// Register all configured runtimes.
+		w.logger.Debug("XXXXXX: Registering configured runtimes...")
 		for _, rt := range commonWorker.GetRuntimes() {
-			if err := w.registerRuntime(rt); err != nil {
+			if err = w.registerRuntime(rt); err != nil {
 				return nil, err
 			}
 		}
+		w.logger.Debug("XXXXXX: Runtime registration complete.")
 
 		// Register transaction scheduler worker role.
-		if err := w.registration.RegisterRole(node.RoleTransactionScheduler,
-			func(n *node.Node) error { return nil }); err != nil {
+		if err = w.registration.RegisterRole(node.RoleTransactionScheduler, func(n *node.Node) error {
+			if w.checkTxEnabled {
+				// Wait until all the runtimes are initialized.
+				w.logger.Debug("XXXXXX: Waiting for runtimes to initialize...")
+				for _, rt := range w.runtimes {
+					select {
+					case <-rt.Initialized():
+					case <-w.ctx.Done():
+						return w.ctx.Err()
+					}
+				}
+				w.logger.Debug("XXXXXX: Runtimes finished initializing.")
+
+				w.logger.Debug("XXXXXX: Obtaining runtime versions...")
+				for _, rt := range n.Runtimes {
+					var grr error
+
+					w.logger.Debug("XXXXXXXXX: Processing runtime...", "rt_id", rt.ID)
+					workerRT := w.runtimes[rt.ID]
+					if workerRT == nil {
+						continue
+					}
+
+					workerHost := workerRT.GetWorkerHost()
+					if workerHost == nil {
+						w.logger.Debug("runtime has shut down",
+							"runtime", rt.ID,
+						)
+						continue
+					}
+					if rt.Capabilities.TEE, grr = workerHost.WaitForCapabilityTEE(w.ctx); grr != nil {
```

Comment:
I found the reason for why SGX tests are not working in this PR. So all batches are being rejected due to the following error:
> roothash/commitment: batch RAK signature invalid

This is because here you actually overwrite the RAK that is registered for the compute runtime (by the compute worker's `RegisterRole` handler). And since each enclave has its own RAK this makes all signatures for the compute enclave invalid.

The solution is either to:
* Support multiple enclaves to be registered for the same runtime in the node descriptor (e.g., multiple RAKs), for different purposes (e.g., one for compute and another one for txnscheduler).
* Remove this for loop so that only the enclave for compute is registered. There is no need to register the txnscheduler's RAK in the registry as this runtime is only used for local checks and it doesn't actually sign any batches.

I would be in favor of option two as it is much simpler and the txnscheduler runtime is only used for CheckTx, which currently doesn't sign the results using RAK. If we later desired to verify that the scheduler actually performed CheckTx, we could do add this later.

Reply:
Ah, that makes sense!  I'll fix it.  Thanks for the help!